### PR TITLE
[TypeDeclaration] Handle return from array_map() on ReturnTypeFromStrictTypedCallRector

### LINKED ISSUE #7827: ReturnTypeFromStrictTypedCallRector turns code into phpstan error
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | [0.15.20](https://github.com/rectorphp/rector/releases/tag/0.15.20)            |

given code 
```php
<?php

final class DemoFile
{
    public function run()
    {
        $range = static function ($start, $end) {
            return array_map(static function ($number) {
                return sprintf('%02d', $number);
            }, range($start, $end));
        };
    }
}

```

is turned into 

https://phpstan.org/r/d11bf4fe-eebc-4ce6-97ec-78e4be083e01

by `ReturnTypeFromStrictTypedCallRector`. 

## Minimal PHP Code Causing Issue

https://getrector.com/demo/11e15340-e07c-4223-a1e7-892c34797dde

## Expected Behaviour

rector should not produce "invalid"/unnecessary code/types

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes https://github.com/rectorphp/rector/issues/7827

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉 /cc @staabm

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

File: rules/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictTypedCallRector.php [hunk 1]
```diff
@@ -116,8 +116,8 @@ public function refactor(Node $node): ?Node
                 return false;
             }
 
-            $currentFunctionLike = $this->betterNodeFinder->findParentType($currentReturn, FunctionLike::class);
-            if ($currentFunctionLike !== $node) {
```
Comment (staabm) [inline rules/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictTypedCallRector.php:120]: I assume the fix for the actual problem is in these lines here.. all other lines changed are only tests and var renames, right?

**Comment by staabm** [comment]
thx for fixing it so fast.

does this also fix
- https://github.com/rectorphp/rector/issues/7827#issuecomment-1463791945
- https://github.com/rectorphp/rector/issues/7827#issuecomment-1463794331
?

if so, would be great to have the test-cases added here

File: rules/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictTypedCallRector.php (hunk 1 above)
Comment (samsonasik) [reply rules/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictTypedCallRector.php:120]: yes, rectify commit are auto-committed due to change from contributors

**Comment by samsonasik** [comment]
I added more fixture:

- https://github.com/rectorphp/rector-src/pull/3471/commits/be5892040b978fb0c905f176d38a53309f5370f7 `str_replace()` can return both string and array, see https://www.php.net/manual/en/function.str-replace.php
- https://github.com/rectorphp/rector-src/pull/3471/commits/361517a51d2ca0cec14bb38d022e47eb738d5b6e , `pathinfo()` can return both string and array, see https://www.php.net/manual/en/function.pathinfo.php

This rule is for strict typed call, so it will got the actual information from the function call definition.

**Comment by staabm** [comment]
> This rule is for strict typed call, so it will got the actual information from the function call definition.

I guess it does currently not take phpstan return type extensions into account.. phpstan itself knows more precise types then the ones rector infers here.

but I think you are saying that it uses native-type knowledge and therefore it makes sense to me - since this means no return-type-extensions involved.

**Comment by TomasVotruba** [comment]
Thank you :+1: Let's ship it

**Comment by staabm** [comment]
awesome, thanks to both of you.

**Pull request merged**